### PR TITLE
Prepare data for REST handlers and construct their gorilla/mux path matchers

### PR DESCRIPTION
**Pull request by vchudnov-g**
This processes HTTP annotations to extract the data that will be needed to generate the REST/proto shim Go source files. As part of that, this parses the URL matching patterns in each HTTP annotation binding, checks for uniqueness of each pattern across all services, and generates the matching URL pattern that will, in subsequent PRs, be passed to gorilla/mux ([GitHub](https://github.com/gorilla/mux), [godoc](https://pkg.go.dev/github.com/gorilla/mux)) as the HTTP server dispatcher.

Note that this does not yet parse the "body" part of HTTP annotations. That will happen in a subsequent PR.

Apologies for this PR being larger than intended: I was iterating to make sure the architecture would be good yet simple enough for our needs.

For reviewing, I suggest following roughly the logical call order, as follows:

- `util/genrest/genserver.go`
- `util/genrest/protomodelcreator.go`
- `util/genrest/protomodel/protomodel.go`
- `util/genrest/internal/pbinfo/pbinfo.go`
- `util/genrest/gomodelcreator.go`
- `util/genrest/gomodel/gomodel.go`
- `util/genrest/gomodel/pathtemplate.go`
- `util/genrest/gomodel/pathtemplateparser.go`
- `util/genrest/gomodel/pathtemplateparser_test.go`
- `util/genrest/gomodel/pathtemplatematch.go`
- `util/genrest/gomodel/pathtemplatematch_test.go`
- `util/genrest/goviewcreator.go`
- `util/genrest/goview/goview.go`
- `util/genrest/goviewcreator_test.go`
- `util/genrest/errorhandling/accumulator.go`
- and then the placeholder generated files in `server/genrest/`

### CODE REVIEW

**Review by software-dov — COMMENTED** [review]
Looks good so far. I'm going to do a second pass to focus on some things I didn't understand first time through, but I also didn't want to make you wait for review comments.

Way, way outside the scope of this review, but I was curious about your decision to write the parser by hand instead of using something like ANTLR.

File: util/genrest/errorhandling/accumulator.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+// AccumulateError stores an error to be reported later.
+func (ea *Accumulator) AccumulateError(err error) {
+	if err == nil {
+		return
+	}
+	if ea.errors == nil {
+		ea.errors = []error{}
+	}
```
[the remark this answers is not in the document]
Comment (vchudnov-g) [reply util/genrest/errorhandling/accumulator.go:35]: Oh, you're right. Thanks!

File: util/genrest/errorhandling/accumulator.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	ea.errors = append(ea.errors, err)
+}
+
+// Error concatenates the string representations of all stored errors and returns it as a single
+// error.
+func (ea *Accumulator) Error() error {
```
[the remark this answers is not in the document]
Comment (vchudnov-g) [reply util/genrest/errorhandling/accumulator.go:41]: Thanks. gapic-generator-go uses a slightly fancier mechanism, so maybe at some point we should switch to using that. But this works for now.

File: util/genrest/gomodel/pathtemplatematch.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+import (
+	"fmt"
+	"strings"
+)
+
+// FindValuesMatching returns the the longest string prefix that matches the templates `first` and
+// `second`, and a bool indicating whether this prefix if a full match for both expressions.
+func FindValuesMatching(first, second PathTemplate) (fullMatch bool, longestMatch string, err error) {
+	one := &traverser{template: first.Flatten()}
+	two := &traverser{template: second.Flatten()}
```
[the remark this answers is not in the document]
Comment (vchudnov-g) [reply util/genrest/gomodel/pathtemplatematch.go:26]: We are passing `one` and `two` to `sementsMatch()`, and they get updated there, so we need to pass by reference. Doing the dereference here rather than at the call site is fairly arbitrary, but reads cleaner to me.

(Separately, `traverser` could have turned out to be heavy weight, and passing by reference would avoid the copying.)

File: util/genrest/gomodel/pathtemplatematch.go [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+	values := make([]string, max(one.Len(), two.Len()))
+	defer func() {
+		longestMatch = strings.Join(values, "")
+	}()
```
[the remark this answers is not in the document]
Comment (vchudnov-g) [reply util/genrest/gomodel/pathtemplatematch.go:31]: correct. I admit it's somewhat misleading to have the explicit returns return empty, but this seems the cleanest way to ensure `longestMatch` is returned as intended.

File: util/genrest/gomodel/pathtemplateparser.go [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+
+// parseLiteral parses a segment with `Kind==Literal`, returning nil if not possible.
+func (parser *Parser) parseLiteral() (*Segment, error) {
+	re := parser.GetRegexp(&parser.reLiteral, "^([a-zA-Z0-9_%]*)")
+	return parser.parseToSegment(re, Literal)
+}
+
+// parseToSegment is a helper function that creates a Segment of the specified kind if the next
+// characters in the parse stream match the expression re.
+func (parser *Parser) parseToSegment(re *regexp.Regexp, kind SegmentKind) (*Segment, error) {
```
[the remark this answers is not in the document]
Comment (vchudnov-g) [reply util/genrest/gomodel/pathtemplateparser.go:157]: It's standard to return `error`, but you're right, since it's never non-`nil` right now. I can omit it. If this ever starts being able to generate an error, I'll add it back.

File: util/genrest/gomodel/pathtemplateparser.go [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+	}
+
+	if !parser.source.ConsumeIf('}') {
+		return segment, fmt.Errorf("expected end-of-variable '}', got %q", parser.source.Str())
+	}
+
+	return segment, nil
+}
+
+// GetRegexp returns the memoized compiled regex corresponding to expr. This assumes the same re is always paired with the same expr.
```
[the remark this answers is not in the document]
Comment (vchudnov-g) [reply util/genrest/gomodel/pathtemplateparser.go:208]: It doesn't _need_ it. It's a (premature?) optimization.

Originally I had each expression being compiled at the start of the function that uses it, but it seemed like compiling the regex repeatedly would be a waste of computation. (I'm assuming it gets recomputed each time, since it is a pointer to a possibly mutable object  returned by a function with constant arguments.)

One way to relax this precondition would be to memoize via `map[string]*regexp.Regexp` instead, assuming map access is cheaper than regexp compilation (which I'm guessing is a reasonable assumption). WDYT?

File: util/genrest/gomodel/pathtemplateparser.go [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+func (parser *Parser) GetRegexp(re **regexp.Regexp, expr string) *regexp.Regexp {
```
[the remark this answers is not in the document]
Comment (vchudnov-g) [reply util/genrest/gomodel/pathtemplateparser.go:209]: I thought about it, but I really like the idea of each regexp being defined near where it's used (since each one is used in one place). Maybe that's too pedantic of me....

File: util/genrest/gomodel/pathtemplateparser_test.go [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+	"testing"
+)
+
+func TestParseTemplate(t *testing.T) {
+	for idx, testCase := range []struct {
+		stringTemplate       string
+		expectSuccess        bool
+		expectParsedTemplate PathTemplate
+	}{
+		// general segment parsing errors
```
[the remark this answers is not in the document]
Comment (vchudnov-g) [reply util/genrest/gomodel/pathtemplateparser_test.go:28]: Oh, good!

**Review by vchudnov-g — COMMENTED** [review]
> I was curious about your decision to write the parser by hand instead of using something like ANTLR.

I thought this was a simple enough case that just coding it by hand would be quicker and simpler to understand, so, TBH, I didn't really consider using pre-existing tools....maybe I should have.

File: util/genrest/gomodel/pathtemplateparser.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (vchudnov-g) [reply util/genrest/gomodel/pathtemplateparser.go]: Changed the memoization to use a more general map. This avoids the precondition, so it's more general and cleaner.

File: util/genrest/errorhandling/accumulator.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (vchudnov-g) [reply util/genrest/errorhandling/accumulator.go:38]: Mainly that we can (and do) embed it in more than one other struct: `gomodel.Model` and `protomodel.Model`

(Even if we were only using it in one place, there's something to be said for splitting the responsibility into a separate data structure.)

File: util/genrest/errorhandling/accumulator.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+	if len(ea.errors) == 0 {
+		return nil
+	}
+	errorStrings := make([]string, len(ea.errors))
+	for idx, err := range ea.errors {
+		errorStrings[idx] = err.Error()
+	}
+	return fmt.Errorf(strings.Join(errorStrings, "\n"))
```
[the remark this answers is not in the document]
Comment (vchudnov-g) [reply util/genrest/errorhandling/accumulator.go:46]: Not quite, because some of errors I'm returning may have newlines themselves; I want to ensure errors are separated by newlines; and I don't want to restrict the use of "[]" at the start/end of errors. (cf https://play.golang.org/p/K4fRZmM26_3).

In addition, this is also less clear IMO than the loop while not saving any computation work, even though it is appealing that it fits in one line of code.

File: util/genrest/gomodel/pathtemplateparser.go [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+//    Segment  = "*" | "**" | LITERAL | Variable ;
+//    Variable = "{" FieldPath [ "=" Segments ] "}" ;
+//    FieldPath = IDENT { "." IDENT } ;
+//    Verb     = ":" LITERAL ;
+// with "**" matching the last part of the path template string except for the Verb.
+func ParseTemplate(template string) (parsed PathTemplate, err error) {
+	return NewParser(template).parse()
+}
+
+////////////////////////////////////////
```
[the remark this answers is not in the document]
Comment (vchudnov-g) [reply util/genrest/gomodel/pathtemplateparser.go:38]: We could. I did debate whether to make this more formal (ie explicitly separating the lexer) right off the bat, but decided to go for something a bit simpler for this specific use case first, while remaining structured. Is this hard to follow? That's good feedback, thanks.

If we were going to make this take a more formal/conventional parsing approach, then yeah, we could do a lexer like in the link you sent or use [Antlr](https://play.golang.org/p/K4fRZmM26_3) like Dov suggested. I can come back to this later.

**Review by software-dov — APPROVED** [review]
Belated approval. I keep attempting to review and getting sucked into minutae, so I've had to heavily structure reading the code and just focus on specific sections.
